### ISSUE TITLE
Rotating around center using Transform3DEffect

### ISSUE THREAD

**Issue by daneshmandi** [Microsoft/Win2D#284]
Hi,

Can you help me with a sample of Transform3DEffect to rotate an image around its center using the following override of DrawImage? 

canvas.DrawImage(transform3DEffect , destinationRect, sourceRect, opacity);

Thanks in advanced!

**Comment by shawnhar** [comment]
Take a look at the Effects Example in Example Gallery.  That rotates around the top left corner of the image.  To rotate around some other point, move the image so the point you want to rotate around becomes the origin, then rotate it, then move it back. eg:

  var matrix = Matrix4x4.CreateTranslation(-imageWidth/2, -imageHeight/2, 0) *
             rotationMatrix *
             Matrix4x4.CreateTranslation(imageWidth/2, imageHeight/2, 0);

**Comment by daneshmandi** [comment]
Hi Shawn,

Thanks for the reply.

Actually I could handle rotating around center point but the problem is the rotated image is clipped when drawn.

Here is my code:

     private void OnDraw(CanvasControl sender, CanvasDrawEventArgs args)
        {
            if (!isReady) return;
            var canvas = args.DrawingSession;
            canvas.Clear(Colors.Transparent);

            var x0 = AssociatedObject.ActualWidth / 2;
            var y0 = AssociatedObject.ActualHeight / 2;

            foreach (var drawingItem in DrawingContext.Items.OrderBy(d => d.ZIndex))
            {
                if (drawingItem.Bitmap != null && drawingItem.Feature != null)
                {
                    var feature = drawingItem.Feature;
                    var centerPoint = new Vector3((float)(x0 + feature.CenterX), (float)(y0 + feature.CenterY), 0);
                    var rotationMatrix = Matrix4x4.CreateRotationZ((float)ConvertToRadians(drawingItem.Rotation), centerPoint);
                    var flipMatrix = Matrix4x4.CreateScale(1, 1, 1, centerPoint);
                    if (drawingItem.IsFilipped)
                    {
                        flipMatrix = Matrix4x4.CreateScale(-1, 1, 1, centerPoint);
                    }
                    rotationMatrix = rotationMatrix * flipMatrix;

                    drawingItem.Transform3DEffect.TransformMatrix = rotationMatrix;

                    Rect sourceRect = new Rect(0, 0, feature.OriginalWidth, feature.OriginalHeight);
                    sourceRect = RotateRect(sourceRect, drawingItem.Rotation);

                    Rect destinationRect = new Rect(x0 + feature.CenterX - (feature.Width / 2), y0 + feature.CenterY - (feature.Height / 2), feature.Width, feature.Height);
                    destinationRect = RotateRect(destinationRect, drawingItem.Rotation);

                    drawingItem.BrightnessEffect.BlackPoint = new Vector2(Math.Max(1 - (float)drawingItem.Brightness, 0), 0);
                    drawingItem.BrightnessEffect.WhitePoint = new Vector2(Math.Min(2 - (float)drawingItem.Brightness, 1), 1);

                    drawingItem.SaturationEffect.Saturation = (float)drawingItem.Saturation;
                    drawingItem.HueRotationEffect.Angle = (float)drawingItem.HueRotation;

                    canvas.DrawImage(drawingItem.HueRotationEffect, destinationRect, sourceRect, (float)drawingItem.Opacity);
                }
            }
        }

     public Point RotatePoint(Point pointToRotate, Point centerPoint, double angleInDegrees)
        {
            double angleInRadians = angleInDegrees * (Math.PI / 180);
            double cosTheta = Math.Cos(angleInRadians);
            double sinTheta = Math.Sin(angleInRadians);
            return new Point
            {
                X =
                    (int)
                    (cosTheta * (pointToRotate.X - centerPoint.X) -
                    sinTheta * (pointToRotate.Y - centerPoint.Y) + centerPoint.X),
                Y =
                    (int)
                    (sinTheta * (pointToRotate.X - centerPoint.X) +
                    cosTheta * (pointToRotate.Y - centerPoint.Y) + centerPoint.Y)
            };
        }

        private Rect RotateRect(Rect rect, double angle)
        {
            Point center = new Point
            {
                X = (rect.Width/2),
                Y = (rect.Height/2)
            };

            Point leftTop = new Point(rect.Left, rect.Top);
            leftTop = RotatePoint(leftTop, center, angle);

            Point rightTop = new Point(rect.Right, rect.Top);
            rightTop = RotatePoint(rightTop, center, angle);

            Point leftBottom = new Point(rect.Left, rect.Bottom);
            leftBottom = RotatePoint(leftBottom, center, angle);

            Point rightBottom = new Point(rect.Right, rect.Bottom);
            rightBottom = RotatePoint(rightBottom, center, angle);

            double minX = Math.Min(Math.Min(leftTop.X, rightTop.X), Math.Min(leftBottom.X, rightBottom.X));

            double minY = Math.Min(Math.Min(leftTop.Y, rightTop.Y), Math.Min(leftBottom.Y, rightBottom.Y));

            double maxX = Math.Max(Math.Max(leftTop.X, rightTop.X), Math.Max(leftBottom.X, rightBottom.X));

            double maxY = Math.Max(Math.Max(leftTop.Y, rightTop.Y), Math.Max(leftBottom.Y, rightBottom.Y));

            return new Rect(minX, minY, (maxX - minX), (maxY - minY));
        }

![rotation](https://cloud.githubusercontent.com/assets/1607430/13237026/7ba8430a-d996-11e5-85aa-787d8d0ce887.jpg)

**Comment by shawnhar** [comment]
I can't tell for sure without being able to run your code, but that sure looks like your source rectangle computation is incorrect.  If you aren't actually intending to crop the image, you could skip all that computation work and just specify source and dest rects expanded by a fixed amount (eg. 2x).

Things I notice about the implementation:

- Rotate rect computes the wrong center point if the rectangle top left is not (0,0)
- There's no need to use Matrix4x4 and Transform3DEffect here - Matrix3x2 and Transform2DEffect can do the same thing more efficiently
- You could simplify RotatePoint a great deal if it took in the same transform matrix that the image effect is using, and used that to transform a Vector2 or Vector3 position value

**Comment by daneshmandi** [comment]
Thanks for the help Shawn.

I have tried your solution to transform image to top-left then applying rotation then transforming it back again to the original location. It worked. You were right there was no need for rotating points and rect. 

I also implemented an Inflate method to expand the Rect to avoid clipping.

Thanks again for this great tool. It is helping me a lot to port [this](http://daneshmandi.com/PortraitPad) to UWP.
